x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/querydsl/query/SingleValueQuery.java

```diff
@@ -168,52 +222,123 @@ public TransportVersion getMinimalSupportedVersion() {
                     DriverContext.WarningsMode.COLLECT,
-                    source.source().getLineNumber(),
-                    source.source().getColumnNumber(),
-                    source.text()
+                    source().source().getLineNumber(),
+                    source().source().getColumnNumber(),
+                    source().text()
                 ),
                 "single-value function encountered multi-value"
             );
             org.apache.lucene.search.Query rewrite = singleValueQuery.rewrite(context.searcher());
             if (rewrite instanceof MatchAllDocsQuery) {
                 // nothing to filter
-                return next.toQuery(context);
+                return next().toQuery(context);
             }
             BooleanQuery.Builder builder = new BooleanQuery.Builder();
-            builder.add(next.toQuery(context), BooleanClause.Occur.FILTER);
+            builder.add(next().toQuery(context), BooleanClause.Occur.FILTER);
             builder.add(rewrite, BooleanClause.Occur.FILTER);
             return builder.build();
         }
 
         @Override
-        protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws IOException {
-            QueryBuilder rewritten = next.rewrite(queryRewriteContext);
-            if (rewritten instanceof MatchNoneQueryBuilder) {
-                return rewritten;
-            }
-            if (rewritten == next) {
-                return this;
-            }
-            return new Builder(rewritten, field, source);
+        protected AbstractBuilder rewrite(QueryBuilder next) {
+            return new Builder(next, field(), source());
+        }
+    }
+
+    /**
+     * Builds a {@code bool} query combining the "next" query, a {@link SingleValueMatchQuery},
+     * and a {@link TermQuery} making sure we didn't ignore any values.
+     * <p>
+     *     This is used in the case when you do {@code text_field == "foo"} and {@code text_field}
+     *     has a {@code keyword} sub-field. See, {@code text} typed fields can't do our equality -
+     *     they only do matching. But {@code keyword} fields *can* do the equality. In this case
+     *     the "next" query is a {@link TermQuery} like {@code text_field.raw:foo}.
+     * </p>
+     * <p>
+     *     But there's a big wrinkle! If you index a field longer than {@code ignore_above} into
+     *     {@code text_field.raw} field then it'll drop its value on the floor. So the
+     *     {@link SingleValueMatchQuery} isn't enough to emulate {@code ==}. You have to remove
+     *     any matches that ignored a field. Luckilly we have {@link IgnoredFieldMapper}! We can
+     *     do a {@link TermQuery} like {@code NOT(_ignored:text_field.raw)} to filter those out.
+     * </p>
+     * <p>
+     *     You may be asking "how would the first {@code text_field.raw:foo} query work if the
+     *     value we're searching for is very long? In that case we never use this query at all.
```

Comment:
> In that case we never use this query at all

Nit: I wonder if this should be a bigger warning at the top of the javdoc. I could imagine somebody (of us) trying to use this for something, and adding a bug because of it 👀 (Not perfect anyway)